src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt

```diff
@@ -0,0 +1,108 @@
+import uk.gov.justice.digital.hmpps.notificationsalertsvsip.dto.visit.scheduler.VisitDto
+import uk.gov.justice.digital.hmpps.notificationsalertsvsip.enums.EmailTemplateNames
+import uk.gov.justice.digital.hmpps.notificationsalertsvsip.enums.VisitEventType
+import uk.gov.justice.digital.hmpps.notificationsalertsvsip.enums.visit.scheduler.VisitRestriction
+import uk.gov.justice.digital.hmpps.notificationsalertsvsip.utils.DateUtils
+import uk.gov.service.notify.NotificationClient
+import java.time.LocalDate
+import java.time.temporal.ChronoUnit
+
+@Service
+class EmailSenderService(
+  @Value("\${notify.email.enabled:}") private val enabled: Boolean,
+  val notificationClient: NotificationClient,
+  val prisonRegisterService: PrisonRegisterService,
+  val prisonerSearchService: PrisonerSearchService,
+  val prisonerContactRegistryService: PrisonerContactRegistryService,
+  val templatesConfig: TemplatesConfig,
+  val dateUtils: DateUtils,
+) {
+  private companion object {
+    val LOG: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun sendEmail(visit: VisitDto, visitEventType: VisitEventType) {
+    if (enabled) {
+      val sendEmailNotificationDto = when (visitEventType) {
+        VisitEventType.BOOKED -> {
+          handleBookedEvent(visit)
+        }
+
+        else -> {
+          LOG.error("visit event type $visitEventType is unsupported")
+          return Unit
+        }
+      }
+
+      notificationClient.sendEmail(
+        templatesConfig.emailTemplates[sendEmailNotificationDto.templateName.name],
+        visit.visitContact.email,
+        sendEmailNotificationDto.templateVars,
+        visit.reference,
+      )
+    } else {
+      LOG.info("Sending Email has been disabled.")
+    }
+  }
+
+  private fun handleBookedEvent(visit: VisitDto): SendEmailNotificationDto {
+    val templateVars = mutableMapOf(
+      "ref number" to visit.reference,
+      "prison" to prisonRegisterService.getPrisonName(visit.prisonCode),
+      "time" to dateUtils.getFormattedTime(visit.startTimestamp.toLocalTime()),
+      "end time" to dateUtils.getFormattedTime(visit.endTimestamp.toLocalTime()),
+      "arrival time" to "45",
+      "dayofweek" to dateUtils.getFormattedDayOfWeek(visit.startTimestamp.toLocalDate()),
+      "date" to dateUtils.getFormattedDate(visit.startTimestamp.toLocalDate()),
+      "main contact name" to visit.visitContact.name,
+      "closed visit" to (visit.visitRestriction == VisitRestriction.CLOSED).toString(),
+      "prisoner" to getPrisoner(visit),
+      "visitors" to getVisitors(visit),
+    )
+    templateVars.putAll(getPrisonContactDetails(visit))
+
+    val templateName = EmailTemplateNames.VISIT_BOOKING
+
+    return SendEmailNotificationDto(templateName = templateName, templateVars = templateVars)
+  }
+
+  private fun getPrisoner(visit: VisitDto): String {
+    return prisonerSearchService.getPrisoner(visit.prisonerId)?.let {
+      it.firstName + " " + it.lastName
+    } ?: "Prisoner"
```

Comment:
suggestion only, can be ignored - could do a toString() like you did below

Reply:
This already returns a string, im not 100% sure what you mean, could you put a code snippet here as example?

Reply:
After discussion i understand overriding the toString() call on the Dto returned is what was suggested. I'll implement this and chase up what the "backup" value should be if the search results in null